src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -85,71 +91,117 @@ class TrimPrimers
     val outSortOrder = sortOrder.getOrElse(in.getFileHeader.getSortOrder)
     val outHeader = in.getFileHeader.clone()
     outHeader.setSortOrder(outSortOrder)
-    val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.queryname, output.toFile, null)
+
+    // Setup the outputs depending on whether or not we have a reference file
+    val (sorter: Option[SortingCollection[SAMRecord]], write: (SAMRecord => Unit), out: SAMFileWriter) = ref match {
+      case Some(path) =>
+        val sorter = buildSorter(SortOrder.coordinate, outHeader)
+        val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.coordinate, output.toFile, null)
+        (Some(sorter), sorter.add _, out)
+      case None =>
+        val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.queryname, output.toFile, null)
+        (None, out.addAlignment _, out)
+    }
```

Comment:
Yeah - I wish it were less so, but I really don't want to force an extra sort if not necessary, and that got me to where I am :(    I'll add more comments.